### PR TITLE
Add stubs for bleach

### PR DESCRIPTION
**Pull request by srittau**
Permission in mozilla/bleach#424

### CODE REVIEW

File: third_party/3/bleach/linkifier.pyi [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from typing import Any, List, MutableMapping, Protocol, Pattern, Iterable, Optional
+
+_Attrs = MutableMapping[Any, str]
+
+class _Callback(Protocol):
+    def __call__(self, attrs: _Attrs, new: bool = ...) -> _Attrs: ...
+
+DEFAULT_CALLBACKS: List[_Callback]
+
+TLDS: List[str]
```
Comment (JelleZijlstra) [inline third_party/3/bleach/linkifier.pyi]: Text actually (the file has unicode_literals). I think most of the strs below should also be Text.

File: third_party/3/bleach/linkifier.pyi [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+def build_url_re(tlds: Iterable[str] = ..., protocols: Iterable[str] = ...) -> Pattern[str]: ...
+
+URL_RE: Pattern[str]
+PHOTO_RE: Pattern[str]
+EMAIL_RE: Pattern[str]
+
+class Linker:
```
Comment (JelleZijlstra) [inline third_party/3/bleach/linkifier.pyi]: Should inherit from object.

However, I just noticed that the stubs are in third_party/3. Why is that? The library looks like it's 2/3-compatible.

File: third_party/3/bleach/sanitizer.pyi [hunk 3]
```diff
@@ -0,0 +1,34 @@
+from typing import List, Dict, Any, Optional, Type, Pattern, Union, Callable, Container
+
+ALLOWED_TAGS: List[str]
```
Comment (JelleZijlstra) [inline third_party/3/bleach/sanitizer.pyi]: Also Text

File: third_party/3/bleach/sanitizer.pyi [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        self,
+        tags: List[str] = ...,
+        attributes: Any = ...,
+        styles: List[str] = ...,
+        protocols: List[str] = ...,
+        strip: bool = ...,
+        strip_comments: bool = ...,
+        filters: Optional[List[Type[_Filter]]] = ...,
+    ) -> None: ...
+    def clean(self, text: str) -> str: ...
```
Comment (JelleZijlstra) [inline third_party/3/bleach/sanitizer.pyi]: Returns Text.

**Comment by srittau** [comment]
I have updated the stubs now to support Python 2. But I wonder whether we should require contributors to provide Python 2 compatible stubs when they submit new stubs. (Of course existing stubs that support Python 2 should continue to do so.) Making stubs Python 2 compatible requires quite a bit of additional work and it's quite possible that a Python 2 version is not that useful anyway, considering Python 2's end of life date.

Personally I suggest, we accept Python 3-only stubs (and also Python 2-only stubs), even if the library supports both versions. Support for the other Python version can always be added later by an interested party.

**Comment by srittau** [comment]
Gentle ping @JelleZijlstra

**Comment by JelleZijlstra** [comment]
Sorry for dropping this!

**Comment by Deimos** [comment]
Just installed the new version of mypy and I'm getting some errors on my codebase now from these stubs. Specifically, I'm using bleach's `Cleaner` class, which has its `__init__` stub as:

```python
class Cleaner(object):
    def __init__(
        self,
        tags: List[Text] = ...,
        attributes: Any = ...,
        styles: List[Text] = ...,
        protocols: List[Text] = ...,
        strip: bool = ...,
        strip_comments: bool = ...,
        filters: Optional[List[Type[_Filter]]] = ...,
    ) -> None: ...
```

I'm passing tuples for both `tags=` and `protocols=`, which works fine but doesn't satisfy these stubs. Maybe the stubs should use `Sequence` instead of explicitly defining `List` (both here and other places where that's relevant)?

Also, I'm passing a list containing a [partial object](https://docs.python.org/3.7/library/functools.html#partial-objects) for `filters=`. Using a partial object this way is shown in the documentation as [a way to both sanitize and linkify in the same step](https://bleach.readthedocs.io/en/latest/linkify.html#using-bleach-linkifier-linkifyfilter). However, the current stubs throw the following error for doing this:

    error: List item 0 has incompatible type "partial[LinkifyFilter]"; expected "Type[Any]"

@srittau would you be able to fix these issues? I could probably also submit a pull request myself to adjust them, if you'd like (though I've never contributed any stubs before so I'm not very familiar with any of what goes into it).

File: third_party/2and3/bleach/__init__.pyi [hunk 5]
```diff
@@ -0,0 +1,29 @@
+from typing import List, Any, Optional, Text
+
+from bleach.linkifier import DEFAULT_CALLBACKS as DEFAULT_CALLBACKS, Linker as Linker
+from bleach.sanitizer import (
+    ALLOWED_ATTRIBUTES as ALLOWED_ATTRIBUTES,
+    ALLOWED_PROTOCOLS as ALLOWED_PROTOCOLS,
+    ALLOWED_STYLES as ALLOWED_STYLES,
+    ALLOWED_TAGS as ALLOWED_TAGS,
+    Cleaner as Clear,
```
Comment (Deimos) [inline third_party/2and3/bleach/__init__.pyi]: Also, I think this was a typo - the name shouldn't be getting changed, right?

**Comment by srittau** [comment]
We welcome pull requests and we can help if you are stuck.

**Pull request merged**